### ISSUE TITLE
np.min is slower than builtin min for small arrays, occasionally VERY slow

### ISSUE THREAD

**Issue by ulfaslak** [numpy/numpy#12350]
Here's a weird bug I just encountered in NumPy. When I run

`%timeit np.min([0.017687381938994294, 0.005285820643230233])`,

in a Jupyter notebook cell, I get vastly different results. Over many tries, the fastest is 18.8 µs and the slowest is a whopping 26.4 ms! Compared with Python's builtin `min` function, which makes the same evaluation in under 300 ns, both extremes are very slow. *Sidenote: running just `np.min(a)` where `a` has been declared in a previous cell, makes no measurable difference.*

In one run I received the warning:

> The slowest run took 31.07 times longer than the fastest. This could mean that an intermediate result is being cached.

Changing the array values did not produce any significant speedups for me.

### Reproducing code example:

Jupyter cell [0]
```python
%timeit np.min([0.017687381938994294, 0.005285820643230233])
# 18.8 µs ± 1.59 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Jupyter cell [1]
```python
%timeit min([0.017687381938994294, 0.005285820643230233])
# 259 ns ± 3.73 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

### Numpy/Python version information:

1.15.1 3.7.0 (default, Jun 28 2018, 07:39:16) 
[Clang 4.0.1 (tags/RELEASE_401/final)]

Python 3.7

**Comment by seberg** [comment]
There is no reason it should even be remotely as fast. Just *converting* the list input to an array must take as long (or longer) then the `min` operation in python. It is not a bug, it is a learning experience for you that numpy in itself is not "making things fast". Numpy is typically fast for medium or large arrays, it tries to be fast for small arrays, but there is only so far you can go.

If you find this is much slower than in previous versions or so, then it might be considered a bug. Sure the different feels large, but for example a quarter of the time is probably already conversion of the list into an array.

**Comment by mattip** [comment]
@seberg I was relating to the secondary comment about "both extremes are very slow", the main issue raised is that `%timeit` in the Jupyter notebook is giving vastly different results when called repeatedly. I think the bug lies outside of numpy, so closing the issue is fine. 

@ulfaslak  if you find the unreliability of very short `%timeit` calls in Jupyeter is numpy-specific, please reopen. You do not say where the notebook kernel is actually running. Are you using a remote machine?

**Comment by seberg** [comment]
I don't think you can reopen if a Member closers, but just comment and we will. I missed the unreliability a bit. But I doubt it is directly numpy related too. There could be other things interfering with the measurement, I guess. CPU clock but also other processes?
Measurements in the millisecond range seems pretty unlikely here, so if that happens reliably only with numpy that would be odd (I cannot reproduce it). `np.min(a)` should make a difference if you have `a = np.asarray([...])` to make sure it already is an array.

**Comment by ulfaslak** [comment]
@seberg I missed an important detail (my mistake). In the sidenote:

> Sidenote: running just np.min(a) where a has been declared in a previous cell, makes no measurable difference.

I meant to say that if *where a has already been declared [as `a = np.array([0.017687381938994294, 0.005285820643230233])`], makes no measurable difference*.

Your points are valid and I cannot guarantee that this is not just a Jupyter issue.

**Comment by seberg** [comment]
@ulfaslak It makes a difference on my machine (maybe up to half of the time), but then I guess I do not have the extreme timings. Will reopen for now, maybe someone on a Mac can try if it is reproducible? I cannot reproduce strong fluctuations on linux.

**Comment by seberg** [comment]
@ulfaslak is there anything special about your system/setup? Is it an Anaconda install or something else? I still doubt this has anything to do with numpy, but if we want to get an idea, probably need to look closer.

**Comment by touqir14** [comment]
Its pretty clear from the benchmarks below that the main overhead is from creating a numpy array inside ```np.min``` : 
```
import numpy as np
x = np.random.rand(10**6)
y = x.tolist()

%timeit np.array(y)
52.1 ms ± 245 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit np.min(y)
52.7 ms ± 175 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit min(y)
17.2 ms ± 175 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit np.min(x)
413 µs ± 1.74 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

```
